code_challenges/code_wars/trigrams.py

```diff
@@ -0,0 +1,19 @@
+# https://www.codewars.com/kata/55d8dc4c8e629e55dc000068
+
+
+def extract_trigrams(words):
+    stopping_point = len(words) - 2
+    if len(words) <= 3:
+        return ""
+    spaces_replaced = ""
+    for i in words:
+        if i == " ":
+            i = "_"
+        spaces_replaced += i
+    words = ""
+    for i in range(len(spaces_replaced)):
+        words += spaces_replaced[i : i + 3] + " "
+        i += 1
+        if i == stopping_point:
+            return words.strip()
+    return words.strip()
```

Comment:
Nice construction of an output string and the use of `.strip`. 

When you are in a for loop were `i` is an index or integer, strongly consider if `i += 1` is needed.  This might lead the reader of the code to think that `i` is incrementing faster than it actually is.  `i` is automatically incremented in the for statement itself, so you are adding one to whatever the value is already incremented to and then it is reassigned that value one line 14 each iteration.

So in this case you are incrementing i only so it will equal the stopping_point at the right time, but stopping_point is a value you chose only for this case so could have been chosen to match what i will be without the additional increment.

For example if you change `stopping_point = len(words) - 2` to be `stopping_point = len(words) - 3` then the `i += 1` line can be removed outright and your code will run the same.

Another thing to consider is 18 is returning early based `i` a count being a certain number.  On line 14 you are creating a loop to do a thing a number of time.  Why not make the number of times on line 14 the actual amount of times you want to do the thing to start with, then you don't need the check at all.  Instead of `len(spaces_replaced)` in `range(len(spaces_replaced))` what value could you use to get the exact right amount of iterations without needing a check and early return?

^ Once this is thought through and implemented, you'll only need line 19s return, as of right now, I don't think line 19 is ever executed.

Reply:
I think `spaces_replaced[i : i + 3]` should be written `spaces_replaced[i:i + 3]` without the spaces between the colon.  But if this was auto-formatted then maybe the rules have changed.

Reply:
I wasn't sure if I did that formatting or not but when I changed it and saved it again, it added the spaces again so it must just be part of the black formatting. 

I can't believe I haven't come across str.replace!!